tasks/broad/ImputationTasks.wdl

```diff
@@ -0,0 +1,761 @@
+    docker: rtidyverse_docker
+    preemptible : 3
+  }
+
+  output {
+    File chunks_info = "~{basename}_chunk_info.tsv"
+    File failed_chunks = "~{basename}_failed_chunks.tsv"
+    Int n_failed_chunks = read_int("n_failed_chunks.txt")
+  }
+}
+
+task MergeImputationQCMetrics {
+  input {
+    Array[File] metrics
+    String basename
+    String rtidyverse_docker
+  }
+
+  Int disk_size = 100 + ceil(size(metrics, "GiB"))
+
+  command <<<
+    Rscript -<< "EOF"
+    library(dplyr)
+    library(readr)
+    library(purrr)
+    library(ggplot2)
+
+    metrics <- list("~{sep='", "' metrics}") %>% map(read_tsv) %>% reduce(`+`) %>% mutate(frac_sites_r2_gt_0.3=total_sites_r2_gt_0.3/total_sites, frac_sites_with_var_r2_gt_0.3=total_sites_r2_gt_0.3/total_sites_with_var)
+
+    write_tsv(metrics, "~{basename}_aggregated_imputation_metrics.tsv")
+    write(metrics %>% pull(frac_sites_with_var_r2_gt_0.3), "frac_well_imputed.txt")
+
+    EOF
+  >>>
+
+  runtime {
+    docker: rtidyverse_docker
+    disks : "local-disk " + disk_size + " HDD"
+    preemptible : 3
+  }
+
+  output {
+    File aggregated_metrics = "~{basename}_aggregated_imputation_metrics.tsv"
+    Float frac_well_imputed = read_float("frac_well_imputed.txt")
+  }
+}
+
+task FailQCThreshold {
+  input {
+    String msg
+  }
+
+  command <<<
+    >&2 echo "QC Failure: ~{msg}"
+    exit 1
+  >>>
+
+  runtime {
+    docker: "ubuntu:18.04"
+  }
+}
+
+task SetIDs {
+  input {
+    File vcf
+    File vcfIndex
+    String output_basename
+    String bcftools_docker
+  }
+
+  Int disk_size = 100 + ceil(2.2 * size(vcf, "GiB"))
+
+  command <<<
+    bcftools annotate ~{vcf} --set-id '%CHROM\:%POS\:%REF\:%FIRST_ALT' -Oz -o ~{output_basename}.vcf.gz
+    bcftools index -t ~{output_basename}.vcf.gz
+  >>>
+
+  runtime {
+    docker: bcftools_docker
+    disks: "local-disk " + disk_size + " HDD"
+    memory: "4 GiB"
+  }
+
+  output {
+    File output_vcf = "~{output_basename}.vcf.gz"
+    File output_vcf_index = "~{output_basename}.vcf.gz.tbi"
+  }
+}
+
+task ExtractIDs {
+  input {
+    File vcf
+    String output_basename
+    Int disk_size = 2*ceil(size(vcf, "GiB")) + 100
+    String bcftools_docker
+  }
+
+  command <<<
+    bcftools query -f "%ID\n" ~{vcf} -o ~{output_basename}.ids
+  >>>
+  output {
+    File ids = "~{output_basename}.ids"
+  }
+  runtime {
+    docker: bcftools_docker
+    disks: "local-disk " + disk_size + " HDD"
+    memory: "4 GiB"
+  }
+}
+
+task SelectVariantsByIds {
+  input {
+    File vcf
+    File ids
+    String basename
+    String gatk_docker
+  }
+
+  Int disk_size = ceil(1.2*size(vcf, "GiB")) + 100
+
+  parameter_meta {
+    vcf: {
+         description: "vcf",
+         localization_optional: true
+       }
+  }
+
+  command <<<
+    cp ~{ids} sites.list
+    gatk SelectVariants -V ~{vcf} --exclude-filtered --keep-ids sites.list -O ~{basename}.vcf.gz
+  >>>
+
+  runtime {
+    docker: gatk_docker
+    disks: "local-disk " + disk_size + " SSD"
+    memory: "16 GiB"
```

Comment:
That's a lot of memory.

Reply:
true, will drop to 4.  Though this task is super quick I think, so really not a spot where there's much savings to be had.